### DISCUSSION TITLE
Ending WPF DataGrid edit when user does not leave row.

### DISCUSSION THREAD

**Discussion by kcabral817** [MarimerLLC/csla#1676] [General]
When using a XamDataGrid, if the user enters data in row but does not leave the row, the edit level for the row will be 2.  This will cause a mismatch when trying to save.  The work around was to force a tab event on the WPF which allows the row to be committed and changes the edit level to match.

           System.Windows.Forms.SendKeys.SendWait("{TAB}");
            System.Windows.Forms.SendKeys.SendWait("+{TAB}");

**Comment by rockfordlhotka** [comment]
The CSLA collection types are tested against the Microsoft data grid controls and how they expect a data source object to implement the data binding interfaces.

Not all third party grids perfectly emulate the Microsoft controls, which can be problematic. 

In short, I'm not surprised you had to do a workaround, but you might also talk to the control vendor to see if they are aware that their behavior isn't matching the "standard" control.

**Reply by rockfordlhotka** [reply]
That said - to my knowledge _all grid controls_ require that focus move out of the row to commit edits to the row. So it might be that this grid is _working exactly right_ and that your UI needs to take steps to ensure the focus moves out of the row when necessary.
